airqo-platform/AirQo-api
# kafka <> device registry

# kafka <> device registry

**_WHAT DOES THIS PR DO?_**

- [x] Integrates Kafka with device registry use cases (activities, devices, sites, airqlouds, locations). That is all CRUD operations.
- [x] Refactors all use cases by reducing number of util files through merging of functionality where scope aligns
- [x] Validates the number of decimal places for the GPS coordinates input in any of the use cases
- [x] Introduces new endpoint/service for consuming Events from Kafka

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
[PLAT-871]

**_WHAT IS THE LINK TO THE PR BRANCH_**
https://github.com/airqo-platform/AirQo-api/tree/ft-kafka-integration-PLAT-871

**_HOW DO I TEST OUT THIS PR?_**

README, development

Ensure that you [have Kafka installed](https://medium.com/@Ankitthakur/apache-kafka-installation-on-mac-using-homebrew-a367cdefd273) and then:

You can utilise Docker solution as demonstrated [here](https://developer.confluent.io/quickstart/kafka-docker/)
 OR open 4 SHELL tabs and:

1. Another for starting Zookeeper (`zookeeper-server-start /usr/local/etc/kafka/zookeeper.properties`)
2. Another for starting Kafka (`kafka-server-start /usr/local/etc/kafka/server.properties`)
3. Another for running  the Kafka Consumer. This is where you will be monitoring the incoming Kafka message for specific topics (`kafka-console-consumer --bootstrap-server localhost:9092 --topic {TOPIC} --from-beginning`)
Topic names have the format: {`DATA_CENTER`}-{`ENV_ACRONYM`}-{`CLASSIFICATION`}-{`ENTITY`}-{`VERSION_NUMBER`} eg: `gcp-dev-createDevice-devices-0`, `gcp-stage-createDevice-devices-0`, `gcp-prod-createDevice-devices-0`
4. And the last one for running the microserivice (`npm run dev-mac`)



**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [x] [create Site](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/create-sites#create-site)
- [x] [Create device](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/create-devices#create-device)
- [x] [Create Location](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/create-locations#create-locations)
- [x] [Create AirQloud](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/airqlouds#create-airqloud)
- [x] [SOFT create Photo](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/photo-registry#soft-create-photo)
- [x] [Deploy device](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/site-activities#deploy-device)
- [x] [Recall device](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/site-activities#recall-device)
- [x] [Maintain device](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/device-registry/site-activities#maintain-device)

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A

**_ARE THERE ANY RELATED PRs?_**
N/A




[PLAT-871]: https://airqoteam.atlassian.net/browse/PLAT-871?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
